### PR TITLE
LF: fix Transaction Decoder/Encoder

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This includes the following changes: 

1. The Decoder was not verifying the value inside a node use the exact
  same version as the node itself.

2. We use the node version directly in the node encoder (instead of
  used `versionedX` node methods).

3. We avoid used of `.asJava` in the encoder.

4. We add test for testing the decode verify 1.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala [hunk 1]
```diff
@@ -115,17 +141,17 @@ object TransactionCoder {
...
-      TransactionOuterClass.KeyWithMaintainers
-        .newBuilder()
-        .setKey(_)
-        .addAllMaintainers(key.maintainers.toSet[String].asJava)
-        .build()
-    )
+    encodeVersionedValue(encodeCid, nodeVersion, key.key).map { encodedKeyValue =>
+      val builder = TransactionOuterClass.KeyWithMaintainers.newBuilder()
+      builder.setKey(encodedKeyValue)
+      key.maintainers.foreach(builder.addMaintainers)
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:152]: :+1:

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala [hunk 2]
```diff
@@ -340,6 +340,161 @@ class TransactionCoderSpec
       }
     }
 
+    def changeVersion(
+        value: com.daml.lf.value.ValueOuterClass.VersionedValue,
+        version: String,
+    ) =
+      value.toBuilder.setVersion(version).build()
+
+    "fail if try to encode a create node containing value with version different from node" in {
```
Comment (cocreature) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala:349]: This sounds like we fail during encoding but that seems to work fine. What fails is decoding. Should we fail to encode such values already?

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala [hunk 3]
```diff
@@ -340,6 +340,161 @@ class TransactionCoderSpec
...
+              ValueCoder.CidDecoder,
+              nodeVersion,
+              encodedNode.toBuilder.setCreate(node).build(),
+            ) shouldBe a[Left[_, _]]
+          )
+        }
+      }
+    }
+
+    "fail if try to encode a fetch node containing value with version different from node" in {
```
Comment (cocreature) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala:391]: same here, we also only fail during decoding

**Review by cocreature — APPROVED** [review]
Good catch! My only question is if we should already fail to encode on version mismatches instead of only failing during decoding.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala:349]: In Scala land the Value in GenNode are not wrapped in VersionedValue any more, so we cannot generate "such" value.

Comment (cocreature) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala:349]: ah I see, makes sense.

**Pull request merged**